content/posts/a_confusing_build_failure.md

```diff
@@ -0,0 +1,122 @@
+
+What's more, we are not listening to the tests. Instead of trying to guess how the library should respond to a request and find ways to mock/stub whatever is required to feed that answer to an assertion, we should add an Adapter layer over the third-party code, test the Adapter with the library and mock callbacks that we may pass to the Adapter layer. 
+
+Perhaps some direct quotes can paint the picture in a better way [ [^2] ]:
+
+> The mess in such tests is telling us that the design isn’t right but, instead of fixing the problem by improving the code, we have to carry the extra complexity in both code and test.
+
+> A second risk is that we have to be sure that the behavior we stub or mock matches what the external library will actually do.
+
+## Summing up
+
+So far we've seen that there is a build failure that was only caught on the CI server - caused by the abuse of mocks on unit tests and the difference in the way CI and local builds are created - which managed to pass to the main trunk and obstruct other pull requests even though no bugs were introduced. Is there a single root cause in this situation? Multiple layers of safety had too be overriden in order for the failure to reach the mainline.
+
+This is not so much an issue of technical deficiency - all developers on the team are quite competent. In order to understand the problems and their various causes let's state the facts and try to analyze the situation.
+
+***Mocks are used excessively and on the wrong context***. They are a special case of test doubles that are used for verification testing [ [^3] ]. 
+
+Using mocks as the de facto test double highlights a lack of understanding of the tools used for testing. Quoting again Nat Pryce and Steve Freeman [ [^1] ]:
+
+> our intention in test-driven development is to use mock objects to bring out relationships between objects.
+
+In every test it should be carefully evaluated what kind of test doubles(if any) are needed and try to minimize mocks only for verifying behavior on interfaces. Of course, there are times where every rule needs to be broken, but most of times we can avoid using tools such as mockito-inline and PowerMock.
+
+***The build failed too late and only on the CI server***. The local workstations and the CI server should build with the same commands in order to avoid unpleasant surprises like this one. Unfortunately, it may be a common assumption that Android Studio builds and tests the same way with the setup of a CI server(this surely caught me by surprise as well). 
+
+Maybe we could check once in a while the way Android Studio builds, no more less because an update can change how things are built and tested. 
```

Comment:
```suggestion
Maybe we could check once in a while the way Android Studio builds, no more no less, because an update can change how things are built and tested. 
```